src/test/tree/DelegationManagerUnit.tree

```diff
@@ -0,0 +1,310 @@
+│                           └── given caller has shares in both EigenPod and StrategyManager
+│                               └── it should set caller delegated to operator, increase operator shares (EPM and SM), call the StakeRegistry, and emit events
+├── when undelegate is called
+│   ├── given caller is not delegated to an operator
+│   │   └── it should revert
+│   ├── given that the caller is registered as operator
+│   │   └── it should revert
+│   ├── given that staker param is zero address
+│   │   └── it should revert
+│   ├── given msg.sender is neither the staker, operator, or delegationApprover (if applicable)
+│   │   └── it should revert
+│   ├── given the msg.sender is the operator or delegationApprover
+│   │   └── it should emit a StakerForceUndelegated event
+│   └── it should emit a StakerUndelegatedEvent and undelegate the staker
+│       ├── given the staker doesn't have delegateable shares
+│       │   └── it should return a zero withdrawal root
+│       └── given the staker has delegateable shares ***
+│           └── it should call _removeSharesAndQueueWithdrawal
+├── when queueWithdrawals is called ***
+│   ├── given that entering the withdrawal queue is paused
+│   │   └── it should revert
+│   └── it should loop through each strategy and call _removeSharesAndQueueWithdrawal
+├── when _removeSharesAndQueueWithdrawal is called
+│   ├── given that the staker is a zero address
+│   │   └── it should revert
+│   ├── given that the length of strategies is 0
+│   │   └── it should revert
+│   └── it should loop through each strategy
+│       ├── given that the staker is delegated to (not zero address)
+│       │   └── it should decrease the operator's shares
+│       ├── given that the strategy is the beacon chain strategy
+│       │   └── it should remove shares from the eigen pod manager
+│       ├── given that the strategy is not the beacon chain eth strategy
+│       │   └── it should remove shares from the strategy manager
+│       ├── given that the staker is delegated to (not zero address)
+│       │   └── it should push a stake update
+│       ├── it should increment the staker's cumulative withdrawals queueWithdrawals
+│       ├── it should calculate and set the withdrawal root as pending
+│       └── it should emit a WithdrawalQueued event and return the withdrawal root
+├── completeQueuedWithdrawal ***
+│   ├── given that the exiting the withdrawal queue is paused
+│   │   └── it should revert
+│   ├── given that the function is reentered
+│   │   └── it should revert
+│   └── it should call _completeQueuedWithdrawal (internal function)
+├── completeQueuedWithdrawals
+│   ├── given that the exiting the withdrawal queue is paused
+│   │   └── it should revert
+│   ├── given that the function is reentered
+│   │   └── it should revert
+│   └── it should loop through each strategy and call _completeQueuedWithdrawal (internal function)
+├── when _completeQueuedWithdrawal is called
```

Comment:
should probably be marked as needing an integration test as well, since so many of the effects are on the EPM / SM (other contracts)

Reply:
Done https://github.com/Layr-Labs/eigenlayer-contracts/pull/332/commits/d9645a9e4a9d32a3526cb5827b8ef30f3d85c5c4